text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+
+## Terminology
+
+* The **Glimmer VM** is the underlying rendering engine which is used by
+  _Ember.js_ and _Glimmer.js_.
+* **Glimmer.js** is a thin wrapper on top of the _Glimmer VM_ which exposes a
+  much simpler API compared to Ember. Historically it has been used to prove out
+  ideas and implementations before bringing them into Ember via RFC, and has
+  been used to write applications which don't require the full feature set of
+  Ember.
+* **Glimmer components** are a newly proposed component API which draw from the
+  experimental APIs provided in Glimmer.js, and Ember.js via
+  [sparkles-component](https://github.com/rwjblue/sparkles-component).
+* **Classic components** refer to the standard component API at the time of this
+  RFC, which have been available in Ember in some form since v1.
+* **Tracked properties** refer to a new method of change tracking which is being
+  proposed in a separate RFC, parallel to this one.
+
+## Motivation
+
+Glimmer components are a simpler base component class that encourage template
+oriented design, thinner class definitions, and functional/composable component
+design. They are easier to understand in isolation, and require less knowledge
+of the framework to use effectively.
+
+This example shows a component written with the classic component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+{{#if (eq this.type 'image'}}
+  <img src={{@post.imageUrl}} title={{@post.imageTitle}}/>
+{{/if}}
+
+{{@post.text}}
+```
+```js
+// components/post.js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['post'],
+  classNameBindings: ['type'],
+  ariaRole: 'region',
+
+  /* Arguments */
+  post: null,
+
+  type: readOnly('post.type'),
+
+  didInsertElement() {
+    if (this.post.type === 'image') {
+      setupImageOverlay(this.element.querySelector('img'));
+    }
+  }
+});
+```
+
+And here is an equivalent component written with the Glimmer component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+<section ...attributes role="region" type={{@post.type}} class="post {{@post.type}}">
+  {{#if (eq @post.type 'image')}}
+    <img {{did-render this.didInsertImage}} src={{@post.imageUrl}} title={{@post.imageTitle}} />
+  {{/if}}
+
+  {{@post.text}}
+</section>
+```
+```js
+// components/post.js
+export default class PostComponent extends GlimmerComponent {
+  didInsertImage(element) {
+    setupImageOverlay(element);
+  }
+}
+```
+
+Glimmer components eliminate many of the common paper cuts that cause confusion
+with classic components, and align more closely with modern template syntax and
+features.
+
+### Outer HTML Semantics
+
+The biggest change Glimmer components make is defaulting to outer HTML
+semantics - What You See Is What You Get, in templates. In the classic component
+API, components had a implicit wrapper element. Given this component template:
+
+```hbs
+Hello, world!
+```
+
+The output by _default_ would be:
+
+```html
+<!-- OUTPUT -->
+<div>
```

Comment:
Worth including the `id="ember-1234"` and `class="ember-view"` gunk here? That adds to the problem, and the overall noisiness, quite a bit.